Exactly Once + Mongo Deletes Support

kurnoolsaketh opened the pull request:
In this PR:
1. fix an issue where repeat document insertions (i.e., documents with colliding `_id`) result in exception looping – this has prevented exactly-once behavior until now.
2. support synchronous deletes in mongoflink. Needed to implement a new `DocumentCodec` and `Document` to achieve this.

link3280 commented:
@kurnoolsaketh Thanks a lot for your PR! The codes generally look good to me. Could you add some tests to verify the deletes?

kurnoolsaketh commented:
> Could you add some tests to verify the deletes? 

yeah definitely – I'll add unit tests to ensure the correct WriteModels are being generated for a given set of Documents.

kurnoolsaketh commented:
@link3280 – addressed your comment and made some changes to a few other tests that were failing due to field ordering being different than expected (for context, when a `CdcDocument` is constructed from a `Document`, the field ordering in the `CdcDocument` is different from the original `Document` since a set is implicitly used in the conversion). This should not matter, as field ordering within a document is not guaranteed in MongoDB without an explicit sort (see: https://jira.mongodb.org/browse/SERVER-50041).

link3280 commented on src/main/java/org/mongoflink/bson/CdcDocument.java:
```diff
@@ -0,0 +1,31 @@
+ * CdcDocument is a bson Document with a delete flag used to indicate if a given Document in a
+ * DocumentBulk represents a mongo delete operation.
+ */
+public class CdcDocument extends Document {
```
Could we use the decorator pattern to avoid creating a new Document? 

Something like:
```
public class CdcDocument extends Document {
    private Document inner;
    private boolean isDelete;
    ....
}
```

kurnoolsaketh replied:
are you suggesting to not invoke the `Document` super constructor?

link3280 replied:
Not really. I mean perhaps we could leave the `CdcDocument` empty (invoke the empty constructor `super()`) and overwrite the methods to use the inner `Document` instead. 

However, I find it cumbersome and hard to maintain because `Document` has a lot of methods. Maybe we could try that approach if reordering becomes a problem. Now I'm good with the PR.

link3280 reviewed (commented):
@kurnoolsaketh Thanks for your update! I left some comments please kindly take a look.
